oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/OpaqueTokenAuthenticationProvider.java

```diff
@@ -0,0 +1,244 @@
+	 * Introspect and validate the opaque
+	 * <a href="https://tools.ietf.org/html/rfc6750#section-1.2" target="_blank">Bearer Token</a>.
+	 *
+	 * @param authentication the authentication request object.
+	 *
+	 * @return A successful authentication
+	 * @throws AuthenticationException if authentication failed for some reason
+	 */
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		if (!(authentication instanceof BearerTokenAuthenticationToken)) {
+			return null;
+		}
+
+		BearerTokenAuthenticationToken bearer = (BearerTokenAuthenticationToken) authentication;
+
+		// introspect
+		HTTPRequest request = introspectionRequest(bearer.getToken());
+		TokenIntrospectionSuccessResponse response = introspectionResponse(request);
+
+		// convert
+		Map<String, Object> claims = response.toJSONObject();
+		if (response.getAudience() != null) {
+			claims.put("aud",
+					Collections.unmodifiableList(response.getAudience().stream()
+							.map(Audience::getValue).collect(Collectors.toList())));
+		}
+		if (response.getClientID() != null) {
+			claims.put("client_id", response.getClientID().getValue());
+		}
+		Instant exp = null;
+		if (response.getExpirationTime() != null) {
+			exp = response.getExpirationTime().toInstant();
+			claims.put("exp", exp);
+		}
+		Instant iat = null;
+		if (response.getIssueTime() != null) {
+			iat = response.getIssueTime().toInstant();
+			claims.put("iat", iat);
+		}
+		if (response.getIssuer() != null) {
+			claims.put("iss", issuer(response.getIssuer().getValue()));
+		}
+		if (response.getNotBeforeTime() != null) {
+			claims.put("nbf", response.getNotBeforeTime().toInstant());
+		}
+		if (response.getScope() != null) {
+			claims.put("scope", Collections.unmodifiableList(response.getScope().toStringList()));
+		}
+
+		// validate
+		// relying solely on the authorization server to validate this token (not checking 'exp', for example)
+		if (!response.isActive()) {
+			throw new OAuth2AuthenticationException(
+					invalidToken("Provided token [" + bearer.getToken() + "] isn't active"));
+		}
+
+		// construct token
+		OAuth2TokenWithClaims withClaims = new OAuth2TokenWithClaims(bearer.getToken(), iat, exp, claims);
+
+		// extract authorities
+		Collection<String> scopes = (Collection<String>) claims.get("scope");
+		List<GrantedAuthority> authorities =
+			Optional.ofNullable(scopes).orElse(Collections.emptyList())
+				.stream()
+				.map(authority -> new SimpleGrantedAuthority("SCOPE_" + authority))
+				.collect(Collectors.toList());
+
+		AbstractAuthenticationToken result = new IntrospectedOAuth2TokenAuthenticationToken(withClaims, authorities);
+		result.setDetails(bearer.getDetails());
+		return result;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public boolean supports(Class<?> authentication) {
+		return BearerTokenAuthenticationToken.class.isAssignableFrom(authentication);
+	}
+
+	private HTTPRequest introspectionRequest(String token) {
```

Comment:
I'd prefer us to use RestTemplate to retrieve and then convert it using Nimbus. This allows for easily modifying timeouts, proxy settings, SSL certificates, etc. Can you please update this to use RestTemplate and allow a user to inject RestOperations?